File: src/cppinterface/cb_minres.jl
@doc raw"""
    CBMinRes(pril::Integer = -1)

default constructor
"""
CBMinRes(pril::Integer = -1) = CBMinRes(@ccall libcb.cb_minres_new(pril::Cint)::Ptr{Cvoid})

@doc raw"""
    cb_set_maxit!(self::CBMinRes, in_maxit::Integer)

set maximum number of iterations
"""
cb_set_maxit!(self::CBMinRes, in_maxit::Integer) = @ccall libcb.cb_minres_set_maxit(self.data::Ptr{Cvoid}, in_maxit::Cint)::Cvoid

@doc raw"""
    cb_get_maxit(self::CBMinRes)

get maximum number of iterations
"""
cb_get_maxit(self::CBMinRes) = @ccall libcb.cb_minres_get_maxit(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_err(self::CBMinRes)

returns the error code of the last call
"""
cb_get_err(self::CBMinRes) = @ccall libcb.cb_minres_get_err(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_nmult(self::CBMinRes)

returns the number of matrix-vector multiplications of the last call
"""
cb_get_nmult(self::CBMinRes) = @ccall libcb.cb_minres_get_nmult(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_residual_norm(self::CBMinRes)

returns the residual norm of last call
"""
cb_get_residual_norm(self::CBMinRes) = @ccall libcb.cb_minres_get_residual_norm(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_avg_reduction(self::CBMinRes)

returns the average of the achieved reduction factor per iteration
"""
cb_get_avg_reduction(self::CBMinRes) = @ccall libcb.cb_minres_get_avg_reduction(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_termprec(self::CBMinRes)

return the (absolute) precision requirement for termination used in the last call
"""
cb_get_termprec(self::CBMinRes) = @ccall libcb.cb_minres_get_termprec(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_compute!(self::CBMinRes, system::CBIterativeSystemObject, x::CBMatrix, termprec::Real, storex::Union{<:CBMatrix,Nothing} = nothing, storestep::Integer = 0)

compute the solution for system into x with (absolute) residual precision termprec
"""
cb_compute!(self::CBMinRes, system::CBIterativeSystemObject, x::CBMatrix, termprec::Real, storex::Union{<:CBMatrix,Nothing} = nothing, storestep::Integer = 0) = @ccall libcb.cb_minres_compute(self.data::Ptr{Cvoid}, system.data::Ptr{Cvoid}, x.data::Ptr{Cvoid}, termprec::Cdouble, (isnothing(storex) ? C_NULL : storex.data)::Ptr{Cvoid}, storestep::Cint)::Cint

@doc raw"""
    cb_set_out!(self::CBMinRes, in_print_level::Integer = 1)

set output stream and level of detail of log output (for debugging)
"""
cb_set_out!(self::CBMinRes, in_print_level::Integer = 1) = @ccall libcb.cb_minres_set_out(self.data::Ptr{Cvoid}, in_print_level::Cint)::Cvoid

File: ConicBundle/cppinterface/cb_minres.cpp
dll void cb_minres_destroy(MinRes* self) {
  delete self;
}

dll MinRes* cb_minres_new(int pril = -1) {
  return new MinRes(&std::cout, pril);
}

dll void cb_minres_set_maxit(MinRes* self, Integer in_maxit) {
  self->set_maxit(in_maxit);
}

dll Integer cb_minres_get_maxit(const MinRes* self) {
  return self->get_maxit();
}

dll int cb_minres_get_err(const MinRes* self) {
  return self->get_err();
}

dll Integer cb_minres_get_nmult(const MinRes* self) {
  return self->get_nmult();
}

dll Real cb_minres_get_residual_norm(const MinRes* self) {
  return self->get_residual_norm();
}

dll Real cb_minres_get_avg_reduction(const MinRes* self) {
  return self->get_avg_reduction();
}

dll Real cb_minres_get_termprec(const MinRes* self) {
  return self->get_termprec();
}

dll int cb_minres_compute(MinRes* self, IterativeSystemObject* system, Matrix* x, Real termprec, Matrix* storex = 0, Integer storestep = 0) {
  return self->compute(*system, *x, termprec, storex, storestep);
}

dll void cb_minres_set_out(MinRes* self, int in_print_level = 1) {
  self->set_out(&std::cout, in_print_level);
}

